### DISCUSSION TITLE
Cannot umount a volume

### DISCUSSION THREAD

**Discussion by myc724** [juicedata/juicefs#228] [Q&A] [answered]
I’m trying to build a flow on the top of JuiceFS and Redis. Everything seems work but I cannot umount volumes at the end. Also, the mountpoints seems created as root while I do not have root access to these machines. Follow lines summarize my flow. Does anyone know what’s the issue? Thanks!

I started a rediss on machine1 for object storage service @ port 6000.
I started a rediss on machine2 for metadata service @ port 7000.

On machine3, I created a volume as follow
    juicefs format --storage=redis --bucket=redis://machine1:6000 redis://machine2:7000/0 test

On machine4 and machine5, I mounted this volume to each machine's local /tmp as follow
    juicefs mount -d redis://machine2:7000/0 /tmp/jfs

Afterward, I can create/write/read files either machine4 or machine5 and JuiceFS synchronized each machine’s local /tmp/jfs perfectly. However, I cannot umount the mountpoint (from either machine4 or machine5). I used this command.
    juicefs --debug umount -f /tmp/jfs

Also, I realized the mountpoint were created as root, while I do not have root access to these machines.
    drwxrwxrwx 4 root        root     4096 Mar  5 18:17 jfs

After I sigkill the “juicefs mount …” processes, mountpioint turned into this.  
    d????????? ? ?           ?           ?            ? jfs

“ls /tmp/jfs” complained as follow.
    ls: cannot access /tmp/jfs: Transport endpoint is not connected

Does anyone know what part were wrong? Thanks!

**Comment by chnliyong** [comment, marked as answer]
I tried mount (the juicefs `main` branch code: a7e2431) with user `ubuntu` in my Ubuntu 20.04 virtual machine:
```
$ ./juicefs mount -d redis://:password@redis.juicefs.local/2 /tmp/jfs
2021/03/08 11:35:56.769313 juicefs[9891] <INFO>: Meta address: redis://:password@redis.juicefs.local/2
2021/03/08 11:35:56.773734 juicefs[9891] <WARNING>: AOF is not enabled, you may lose data if Redis is not shutdown properly.
2021/03/08 11:35:56.775340 juicefs[9891] <INFO>: Data use s3://juicefs-dev/dev/
2021/03/08 11:35:56.775496 juicefs[9891] <INFO>: Cache: /var/jfsCache/c21e50f9-2997-4e6a-bbd8-149ca962d338 capacity: 1024 MB
2021/03/08 11:35:56.775640 juicefs[9891] <WARNING>: statfs /var/jfsCache/c21e50f9-2997-4e6a-bbd8-149ca962d338/: permission denied
2021/03/08 11:35:56.776179 juicefs[9891] <WARNING>: statfs /var/jfsCache/c21e50f9-2997-4e6a-bbd8-149ca962d338/: permission denied
2021/03/08 11:35:56.928953 juicefs[9891] <INFO>: OK, dev is ready at /tmp/jfs

$ df -h | grep JuiceFS
JuiceFS:dev     1.0P   64K  1.0P   1% /tmp/jfs

$ ls -ld /tmp/jfs
drwxrwxrwx 1 root root 4096 Mar  3 09:09 /tmp/jfs

$ ./juicefs umount /tmp/jfs
<umount success, no output>

$ ls -ld /tmp/jfs
drwxrwxr-x 2 ubuntu ubuntu 4096 Mar  8 11:32 /tmp/jfs

$ df -h | grep JuiceFS
<no output>
```

I cannot reproduce your problem, can you provide more information like operation system ( `cat /etc/os-release; uname -a` ), JuiceFS version (`./juicefs -V`) etc. ?

**Reply by myc724** [reply]
juicefs version 0.11.0 (2021-03-01T08:34:01Z a354b98).
I’ve tried compiled-from-source-code / downloaded-binary and got the same result. 
I'm using a heterogeneous system: enclosed “cat /etc/os-release; uname -a” from 2 machines.

Your log also showed that /tmp/jfs was mounted as root.
The flow (that I’m building) can not run as root, aka. the flow needs to proceed JuiceFS format/mount/umount as regular user.
Is this doable? Thanks!

=====

cat /etc/os-release; uname -a
NAME="CentOS Linux"
VERSION="7 (Core)"
ID="centos"
ID_LIKE="rhel fedora"
VERSION_ID="7"
PRETTY_NAME="CentOS Linux 7 (Core)"
ANSI_COLOR="0;31"
CPE_NAME="cpe:/o:centos:centos:7"
HOME_URL="https://www.centos.org/"
BUG_REPORT_URL="https://bugs.centos.org/"
CENTOS_MANTISBT_PROJECT="CentOS-7"
CENTOS_MANTISBT_PROJECT_VERSION="7"
REDHAT_SUPPORT_PRODUCT="centos"
REDHAT_SUPPORT_PRODUCT_VERSION="7"

Linux machine1 3.10.0-1127.10.1.el7.x86_64 #1 SMP Wed Jun 3 14:28:03 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux

===== 

NAME="CentOS Linux"
VERSION="7 (Core)"
ID="centos"
ID_LIKE="rhel fedora"
VERSION_ID="7"
PRETTY_NAME="CentOS Linux 7 (Core)"
ANSI_COLOR="0;31"
CPE_NAME="cpe:/o:centos:centos:7"
HOME_URL="https://www.centos.org/"
BUG_REPORT_URL="https://bugs.centos.org/"

CENTOS_MANTISBT_PROJECT="CentOS-7"
CENTOS_MANTISBT_PROJECT_VERSION="7"
REDHAT_SUPPORT_PRODUCT="centos"
REDHAT_SUPPORT_PRODUCT_VERSION="7"

Linux machine2 3.10.0-1062.12.1.el7.x86_64 #1 SMP Tue Feb 4 23:02:59 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux

**Reply by chnliyong** [reply]
I created an CentOS 7.2 public cloud virtual machine and reproduced your problem.
In short, the `umount` is too old to support unmount the _fuse_ filesystem, and we can use `fusermount -u /tmp/jfs` to unmount the _fuse_ filesystem. 

It seems the `umount`(in package `util-linux`) is not support _fuse_ mount. I read the `util-linux` source code commit history, the first time it add support for _fuse_ mounts is [this commit](https://git.kernel.org/pub/scm/utils/util-linux/util-linux.git/commit/?id=5fea669e9ef0a08804f72bb40f859f239f68c30a), which happened on the day **2018-11-30** . On the server I reproduced this problem, the version is `2.23.2` as show below:
```
$ rpm -qif /bin/umount
Name        : util-linux
Version     : 2.23.2
Release     : 33.el7_3.2
Architecture: x86_64
Install Date: Thu 17 Aug 2017 07:45:12 AM CST
Group       : System Environment/Base
Size        : 8192921
License     : GPLv2 and GPLv2+ and LGPLv2+ and BSD with advertising and Public Domain
Signature   : RSA/SHA256, Thu 13 Apr 2017 09:40:56 AM CST, Key ID 24c6a8a7f4a80eb5
Source RPM  : util-linux-2.23.2-33.el7_3.2.src.rpm
Build Date  : Thu 13 Apr 2017 04:43:54 AM CST
Build Host  : c1bm.rdu2.centos.org
Relocations : (not relocatable)
Packager    : CentOS BuildSystem <http://bugs.centos.org>
Vendor      : CentOS
URL         : http://en.wikipedia.org/wiki/Util-linux
Summary     : A collection of basic system utilities
Description :
The util-linux package contains a large variety of low-level system
utilities that are necessary for a Linux system to function. Among
others, Util-linux contains the fdisk configuration tool and the login
program.
```
The release date of this version is **2013-07-31** from [this commit](https://git.kernel.org/pub/scm/utils/util-linux/util-linux.git/commit/?h=stable/v2.23&id=e6a72859700384a7e0565472ecf5c48a6392be0e), obviously not support unmount the _fuse_ filesystem.

**Reply by chnliyong** [reply]
If you mount JuiceFS with regular user, please use `--cache-dir` to change the default cache directory (`/var/jfsCache`), otherwise it will continuously complain the `no permission` warning log.

**Reply by chnliyong** [reply]
Will fix in this PR #242

**Reply by myc724** [reply]
"fusermount -u <mountpoint>" works. Thanks a lot!
Look for PR #242.

**Reply by myc724** [reply]
Thanks a lot for the quick fix!
